src/query/api/v1/handler/prometheus/native/read.go

```diff
@@ -84,13 +85,26 @@ func NewPromReadHandler(
 }
 
 func (h *PromReadHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	result, params, err := h.ServeHTTPWithEngine(w, r, h.engine)
+	if err != nil {
+		return
+	}
+
+	// TODO: Support multiple result types
+	w.Header().Set("Content-Type", "application/json")
+	w.Header().Set("Access-Control-Allow-Origin", "*")
+	renderResultsJSON(w, result, params)
+}
+
+// ServeHTTPWithEngine returns query results from the storage
+func (h *PromReadHandler) ServeHTTPWithEngine(w http.ResponseWriter, r *http.Request, engine *executor.Engine) ([]*ts.Series, models.RequestParams, error) {
 	ctx := context.WithValue(r.Context(), handler.HeaderKey, r.Header)
 	logger := logging.WithContext(ctx)
 
 	params, rErr := parseParams(r)
 	if rErr != nil {
 		xhttp.Error(w, rErr.Inner(), rErr.Code())
-		return
+		return nil, emptyReqParams, rErr
```

Comment:
Will we end up calling `xhttp.Error` twice here? (looks like we call it again in the caller). Maybe we can just render results + errors in the caller. That would also eliminate the weird dichotomy between writing errors here but returning results.

Reply:
Hmm not quite sure I understand, you mean move the `xhttp.Error` code to `ServeHTTP`?